### PR TITLE
Made changes to `HttpTrait` and added `ProxyTrait`.

### PR DESCRIPTION
**Pull request by vcolin7**
- Changes in `HttpTrait`:
  - Renamed `pipeline()` to `httpPipeline()`
  - Renamed `addPolicy()` to `addHttpPipelinePolicy()`
  - Renamed `retryOptions()` to `httpRetryOptions()`
  - Added `httpRedirectOptions()`
- Added `ProxyTrait` with `proxyOptions()` in it

### CODE REVIEW

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java [hunk 1]
```diff
@@ -5,17 +5,18 @@
...
+import com.generic.core.http.models.HttpRedirectOptions;
 import com.generic.core.http.models.HttpRetryOptions;
 import com.generic.core.http.pipeline.HttpPipeline;
 import com.generic.core.http.pipeline.HttpPipelinePolicy;
 
 /**
- * An {@link com.generic.core.models.traits SDK for Java trait} providing a consistent interface for configuration
- * of HTTP-specific settings. Refer to the SDK for Java
+ * An {@link com.generic.core.models.traits SDK for Java trait} providing a consistent interface for configuration of
+ * HTTP-specific settings. Refer to the SDK for Java
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java:15]: Wording is a bit confusing here, and likely the same in `azure-core` too.

```suggestion
 * A {@link com.generic.core.models.traits trait} providing a consistent interface for configuration of
 * HTTP-specific settings. Refer to the SDK for Java
```

I don't think we need to, or should, include `SDK for Java`, it's a bit redundant.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java [hunk 2]
```diff
@@ -43,12 +45,12 @@ public interface HttpTrait<T extends HttpTrait<T>> {
...
      * documentation of types that implement this trait to understand the full set of implications.</p>
      *
-     * @param httpClient The {@link HttpClient} to use for requests.
+     * @param client The {@link HttpClient} to use for requests.
      *
      * @return Returns the same concrete type with the appropriate properties updated, to allow for fluent chaining of
      * operations.
      */
-    T httpClient(HttpClient httpClient);
+    T httpClient(HttpClient client);
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java:53]: Why for everything else did we match name for name and this we reduce it?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * <a href="https://aka.ms/azsdk/java/docs/http-client-pipeline">HTTP clients and pipelines</a> documentation for more
+ * details on proper usage and configuration of the SDK for Javan HTTP clients.
+ *
+ * @param <T> The concrete type that implements the trait. This is required so that fluent operations can continue to
+ * return the concrete type, rather than the trait type.
+ *
+ * @see com.generic.core.models.traits
+ * @see ProxyOptions
+ */
+public interface ProxyTrait <T extends HttpTrait<T>> {
```
Comment (alzimmermsft) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java:21]: Is this a typo?

```suggestion
public interface ProxyTrait <T extends ProxyTrait<T>> {
```

Or are we tying `ProxyOptions` to HTTP only?

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+package com.generic.core.models.traits;
+
+import com.generic.core.http.client.HttpClient;
+import com.generic.core.http.models.ProxyOptions;
+
+/**
+ * An {@link com.generic.core.models.traits SDK for Java trait} providing a consistent interface for configuration
+ * of proxy-specific settings. Refer to the SDK for Java
+ * <a href="https://aka.ms/azsdk/java/docs/http-client-pipeline">HTTP clients and pipelines</a> documentation for more
```
Comment (srnagar) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java:12]: Proxy options are not specific to HTTP. So, all documentation should be updated to not include HTTP-specific details.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java (hunk 3 above)
Comment (srnagar) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java:21]: This should be `ProxyTrait` and not `HttpTrait`.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java [hunk 5]
```diff
@@ -115,11 +117,28 @@ public interface HttpTrait<T extends HttpTrait<T>> {
...
+     * based on the settings provided to this trait. Additionally, there may be other APIs in types that implement this
+     * trait that are also ignored if an {@link HttpPipeline} is specified, so please be sure to refer to the
+     * documentation of types that implement this trait to understand the full set of implications.</p>
+     *
+     * @param httpRedirectOptions The {@link HttpRedirectOptions} to use for all the requests made through the client.
+     *
+     * @return Returns the same concrete type with the appropriate properties updated, to allow for fluent chaining of
+     * operations.
+     */
+    T httpRedirectOptions(HttpRedirectOptions httpRedirectOptions);
```
Comment (srnagar) [inline sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java:143]: This is something we should discuss adding to azure-core as well. Redirect options are generally not needed on every client builder. So, I am inclined to not add this in the trait interface and let users use the addPolicy method to add the redirect policy.

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java (hunk 3 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/ProxyTrait.java:21]: It's a typo, thanks for pointing it out :)

File: sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java (hunk 2 above)
Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java:53]: The pipeline method shows `pipeline` as well. I think I'll change both to match the name of the type.

Comment (vcolin7) [reply sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/models/traits/HttpTrait.java:53]: We don't use the `http` prefix in any other parameter name in this interface.

Also related, I made a mistake when I added `httpRedirectOptions(HttpRedirectOptions httpRedirectOptions)` so I'm going to remove the prefix from the param name there as well.

**Review by samvaity — APPROVED** [review]

**Review by alzimmermsft — APPROVED** [review]

**Review by JonathanGiles — APPROVED** [review]

**Review by srnagar — APPROVED** [review]
LGTM

**Pull request merged**